aten/src/ATen/native/TensorShape.cpp

```diff
@@ -50,8 +50,30 @@ static void check_cat_no_zero_dim(TensorList tensors) {
 Tensor & cat_out(Tensor & result, TensorList tensors, int64_t dim) {
   check_cat_no_zero_dim(tensors);
   dim = legacy_cat_wrap_dim(dim, tensors);
-  return at::_cat_out(result, tensors, dim);
+#ifdef BUILD_NAMEDTENSOR
+  auto outnames = namedinference::compute_cat_outnames(tensors);
+  {
+    NoNamesGuard guard;
+#endif
+    at::_cat_out(result, tensors, dim);
+#ifdef BUILD_NAMEDTENSOR
+  }
+  namedinference::propagate_names(result, std::move(outnames), /*validate_names=*/false);
+#endif
+  return result;
+}
+
+#ifdef BUILD_NAMEDTENSOR
+Tensor& cat_out(Tensor& result, TensorList tensors, Dimname dim) {
+  TORCH_CHECK(!tensors.empty(), "expected a non-empty list of Tensors");
+  return at::cat_out(result, tensors, dimname_to_position(tensors[0], dim));
```

Comment:
Not sure if this is correct but I'm thinking that if the passed in `dim` refers to a specific tag, and tensors[0]'s dimnames are tagged differently than tensors[1]'s (e.g. A.x, A.y vs A.y, A.x), then should that be an error? 

I think in this case if `dim==A.x` it would find the first dimension in the first tensor, and A.x would be compatible with A.y, but since the tag was explicit maybe shouldn't be? or maybe I'm wrong and A.x is already incompatible with A.y? Regardless a test with tags might be helpful to clarify what's expected.

Reply:
We're not going to include tagged names in the first iteration of named tensors. Eventually, before the release, I will send a PR to disable their creation.

For context: I've been thinking about tagged names a lot and I think that they are most likely a bad idea and might not exist in a future iteration of named tensors; as you pointed out, there are a lot of edge cases with tags.

Reply:
ah OK, would like to see more on the thinking there and what we're going to do. It seems like a meaningful use case and we do already have a lot of code for handling them, but yes it does seem pretty complicated.